synapse/storage/databases/main/user_directory.py

```diff
@@ -369,6 +371,22 @@ def _get_next_batch(txn: LoggingTransaction) -> Optional[List[str]]:
 
         return len(users_to_work_on)
 
+    async def is_excluded_from_user_dir(self, user: str) -> bool:
+        """Certain classes of local user are omitted from the user directory.
+        Is this user one of them?
+        """
+        if self.hs.is_mine_id(user):
+            # Support users are for diagnostics and should not appear in the user directory.
+            if await self.is_support_user(user):
+                return True
+            # Deactivated users aren't contactable, so should not appear in the user directory.
+            if await self.get_user_deactivated_status(user):
+                return True
+            # App service users aren't usually contactable, so exclude them.
+            if self.get_if_app_services_interested_in_user(user):
+                return True
```

Comment:
wonder if we should do this test first, since it doesn't require a database hit and should be pretty quick.